### ISSUE TITLE
Can't combine Duration.Finite with a Duration

### ISSUE THREAD

**Issue by CptnKirk** [zio/zio#3011]
Regression from RC17 -> RC18

It's not obvious how to combine a `Duration.Finite` and a `Duration` from `.timed`.

As a result a `mapAccum(Duration.Zero){...}` from RC17 breaks in RC18.

**Comment by adamgfraser** [comment]
Let me take a look.

**Comment by adamgfraser** [comment]
@CptnKirk I don't think there is an issue in combining `Duration.Finite` and `Duration`. You can just do it with `+` assuming you want to add them. I think the issue is that `Duration.Zero` returns a `Duration.Finite` so when you do `mapAccum(Duration.Zero)(???)` the inferred type of the state being accumulated is `Duration.Finite` instead of `Duration` and so `+` doesn't return the right type (since adding two finite duration could potentially return an infinite duration given overflow).

You can resolve by using a type ascription (e.g. `mapAccum(Duration.Zero: Duration)`) or by using a type annotation. Note that this isn't really a ZIO specific issue. You have the problem in Scala in general if you try to do something like `as.foldRight(Nil)((_ :: _)`.

We could potentially solve by having `Duration.Zero` return `Duration` instead of `Duration.Finite` but we are throwing away information and there are other cases where we want to know that the duration is finite. We could also add a smart constructor `val zero: Duration = Duration.Zero` but that still only addresses this specific case (e.g. `Duration.fromNanos` also returns a finite duration for example).

**Comment by CptnKirk** [comment]
I see.  I was confused by the error message and pointer at the `+`
```
[error]  found   : zio.duration.Duration
[error]  required: zio.duration.Duration.Finite
[error]      .mapAccum(Duration.Zero){case (time, el) => (time + el._3, (el, time + el._3))}
[error]                                                        ^
```

Adding type ascription did solve my problem.  However, this isn't very ergonomic.  In fact it seems that combining Finite values results in a Duration.  This is also unexpected, as two finite values can't combine to form an infinite (only a larger finite value).  You mentioned something about an overflow, however I'd expect that this would be modeled differently.  Or you could back the scalar value of the duration implementation with a BigInt or some non-overflowable type.

I'd also expect the return values of `.timed` to be Finite, since any returning timed value must be a discrete finite value.  Having `.timed` return non-Finite Durations seems to communicate information that isn't possible. 

If timers returned Finite durations and combining Finite durations also yielded Finite durations, then we could fold/aggregate naturally without ascriptions.

**Comment by adamgfraser** [comment]
Resolved by #3032.
